staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_discovery.go

```diff
@@ -0,0 +1,147 @@
+	// +optional
+	Status APIGroupDiscoveryStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// APIVersionDiscovery holds a list of APIResourceDiscovery types that are served for a particular version within an API Group.
+type APIVersionDiscovery struct {
+	// Version is the name of the version within a group version.
+	Version string `json:"version" protobuf:"bytes,1,opt,name=version"`
+	// Resources is a list of APIResourceDiscovery objects for the corresponding group version.
+	// +listType=map
+	// +listMapKey=resource
+	Resources []APIResourceDiscovery `json:"resources,omitempty" protobuf:"bytes,2,rep,name=resources"`
+}
+
+// APIResourceDiscovery provides information about an API resource for discovery.
+type APIResourceDiscovery struct {
+	// Resource is the plural name of the resource.  This is used in the URL path and is the unique identifier
+	// for this resource across all versions.
+	// resources with non-"" groups are located at /apis/<APIGroupDiscovery.objectMeta.name>/<APIVersionDiscovery.version>/<APIResourceDiscovery.Resource>
+	// resource with "" groups are located at /api/v1/<APIResourceDiscovery.Resource>
+	Resource string `json:"resource" protobuf:"bytes,1,opt,name=resource"`
+	// ReturnType describes the type of serialization that will be returned from this endpoint.
+	ReturnType APIDiscoveryKind `json:"returnType" protobuf:"bytes,2,opt,name=returnType"`
+	// Scope indicates the scope of a resource, either Cluster or Namespace
+	Scope ResourceScope `json:"scope" protobuf:"bytes,3,opt,name=scope"`
+	// SingularName is the singular name of the resource.  This allows clients to handle plural and singular opaquely.
+	// The singularName is more correct for reporting status on a single item and both singular and plural are allowed
+	// from the kubectl CLI interface.
+	SingularName string `json:"singularName" protobuf:"bytes,4,opt,name=singularName"`
+	// Verbs is a list of supported kube verbs (this includes get, list, watch, create,
+	// update, patch, delete, deletecollection, other)
+	Verbs Verbs `json:"verbs" protobuf:"bytes,5,opt,name=verbs"`
+	// ShortNames is a list of suggested short names of the resource.
+	// +listType=set
+	ShortNames []string `json:"shortNames,omitempty" protobuf:"bytes,6,rep,name=shortNames"`
+	// Categories is a list of the grouped resources this resource belongs to (e.g. 'all')
+	// +listType=set
+	Categories []string `json:"categories,omitempty" protobuf:"bytes,7,rep,name=categories"`
+	// Subresources is a list of subresources provided by this resource. Subresources are located at /apis/<APIGroupDiscovery.objectMeta.name>/<APIVersionDiscovery.version>/<APIResourceDiscovery.Resource>/name-of-instance/<APIResourceDiscovery.subresources[i].subresource>
+	// +listType=map
+	// +listMapKey=subresource
+	Subresources []APISubresourceDiscovery `json:"subresources,omitempty" protobuf:"bytes,8,rep,name=subresources"`
```

Comment:
Is it actually possible to have a subresource without a parent? I think we have some assumptions baked into our apiserver that looks at the parent resource for determining a subresource's scope (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L220-L230) and

Reply:
I think I agree that we don't have a great way to express that, but going back to your point below:
> maintaining parallel versions... and status.staleVersions lists callers have to correlate themselves seems weird for an API that is effectively all status
Having the subresources under the parent resources does automatically correlate them. On the other hand, the empty resource seems hypothetical so I'm on the fence here.

Reply:
Given the manner in which the URL is built, there is always a logical resource.

Reply:
> allows servers to advertise a subresource without a dedicated parent resource

You can list whatever, but it's impossible to make a request without putting something into the "resource" path element. I think such a thing is already broken.

Reply:
It's impossible to have a sub resource without a kind, but not impossible to not have a subresource and a resource and no kind.  We should not prohibit that.

Reply:
> It's impossible to have a sub resource without a kind, but not impossible to
> 
> * have a subresource
> * have a parent resource
> * not have a kind for the parent resource
> 
> We cannot introduce an API that can't express things that previous discovery does.
> 
> EDIT: Reading this carefully, I see how we got to this point, and my previous comments seem to be covered in the context of "we are trying to force fit hierarchal resource retrieval into a single object". Removing the comment, and leaving "we need to be able to represent all things previous discovery represented". Doing otherwise breaks a v1 API.

This can still be expressed in the new API though? We are creating a hierarchical representation with the new API but the situation of 

> have a subresource
> have a parent resource
> not have a kind for the parent resource

can still be represented with a subresource nested within a parent resource without a kind.

Reply:
> Doing otherwise breaks a v1 API.

No, we are not removing or deprecating the v1 API, we are adding a v2 API.

Reply:
Since the pitch is that it makes clients better, if you ALSO change that API so that old clients have to reason about the set of changes, I think that's not accomplishing the goal the KEP described:

> All clients of Kubernetes APIs must send a request to every
group-version in order to "discover" the available APIs. This causes a
storm of requests for clusters and is a source of latency and
throttling. This KEP proposes centralizing the "discovery" mechanism
into one document so that clients only need to make one request to the
API server to retrieve all the operations available.

"All clients" means (to me) that it should be trivial for all clients to switch to this new endpoint and not have to reason about "some things that could be represented before that I worked with no longer work".

Reply:
I think the goal is for there to be minimal disruptive changes, [improve upon the lessons learned from the first API](https://github.com/kubernetes/kubernetes/pull/111409#issuecomment-1198536637), and we'll offer a conversion for existing clients in our existing rest mapper client-go code so so the change is almost transparent for go clients (kubectl, controllers, etc are among the biggest beneficiaries of this change).

Reply:
No matter what we do, clients will have to adjust to new endpoint semantics.  Given that clients will have to adjust regardless, improving the API shape during that adjustment doesn't make it substantively worse for a client (these changes fix logical inconsistencies the discovery we have), and there is a clear benefit to a client to make this change, so adoption is incentivized.

There is not a way to create a new endpoint that behaves differently without having clients adjust.

Reply:
> (these changes fix logical inconsistencies the discovery we have)

This sounds like "things that could previously be described cannot be", and that sounds like "deprecation and removal of a behavior from our v1 API", which is not what I saw described in the KEP.  Apologize if I've missed a detailed list of the inconsistencies and how we believe they don't impact users in the KEP, but at least based on the reading the bar here is "move our v1 client to this new format for performance reasons without regressing users".

If it's important to remove those logical inconsistencies, we should do that via a new KEP that rejects those discovery documents during APIService registration, and it should go through the normal API change process to do so, and if we did find users using it we would need to establish a workaround for deprecation or potentially declaring that supported and not proceeding further with removal.

The default expectation of API change in general and this API as described in the KEP specifically seems to be "represent everything our previous discovery represents".

Reply:
Because subresources fulfill a URL pattern that is deeper than resource, it is logically impossible to have functional subresource that does not have a parent resource because no server could interpret it.  A subresource-only in discovery implies the name of its parent resource. The new schema represents the same information via nesting, so they are fully convertible back and forth.

Reply:
Maybe we should call this new thing "cachedDiscovery" instead of "discovery v2". Would that make it clearer that v1 is never going away, and clients may use whichever suits them better?

Reply:
> Because subresources fulfill a URL pattern that is deeper than resource, it is logically impossible to have functional subresource that does not have a parent resource because no server could interpret it. A subresource-only in discovery implies the name of its parent resource. The new schema represents the same information via nesting, so they are fully convertible back and forth.

It is logically possible today to have subresources with a parent resource where the parent resource has no registered Kind (which is the criticism of the API in the comments right now, not about having no parent resource).

If we are proposing to null out the Kind for those parent resources and other "unknown fields", then the next point is to critiique whether we want the synthetic parent object to have empty fields, pointers, or pick a serialization that doesn't require those synthetic fields to be wierdly empty.  I don't like wierdly empty fields in general, since now a client has to handle those.  Clients should not have to handle guessing whether the parent object is real or virtual (has subresources, parent resource, but no parent kind or parent operations) or not in the shape of our new API.

Reply:
Ok, so I did some spelunking in the installer.go changes and how we got there, and I got a bit sad.

[61ba8ca0bc072035a5fd15301854b23eabec1e70 (2017)](https://github.com/kubernetes/kubernetes/pull/49357) made it so that you could no longer register a subresource - we did a lookup of the kind so that we could fill out a few options more cleanly in the docs (the doc message, and the combined name).  So we've prevented this for a very long time, and the various cleanups David did shuffled some of the code but didn't change that (and that's why he was confident).  I reviewed this but didn't catch that restriction - I would have preferred that restriction not land, but it's in and done for a long time now.

So anyone implementing an aggregated API with Kube code would be bound by this - I think it's unlikely we have lots of very old apiservers, so we can exclude those.  What I can't exclude without more data is people building aggregated API servers outside of the Go ecosystem - I have no direct evidence of it, but any code restriction would not apply (and of course they'd have to at minimum emulate the discovery doc, which is not hard).

---

For that reason, I'd prefer making the parent resource definition in the API (everything except the name) be optional.  The name is the minimum required element in an apiresource, and only implementation constrains this setting.  Putting the other attributes tied to the parent resource in an optional struct pointer block would sufficiently capture this (you can have a sub resource without a parent resource kind, but not the name).  

A second factor would be that today in the discovery list you can have conflicting values for resources (namespace scope and such), and I'd want us to explicitly describe in the KEP and API doc how that conflict would be handled (subresource A declared as namespaced, subresource B is declared as not).  In general defensiveness should be explicitly handled as we reduce these.

Reply:
Second note, looking at usage in the ecosystem was valuable - after reviewing the code users are writing to resolve this, the structure David is proposing reduces the rist of incorrect assumptions at the client through structure, and gives us the opportunity to defend on the server side users from impossible results.  That (to me) justifies the introduction of the new structure, as long as we represent previously legal things.  The alternate - an entrypoint returning the original objects - would be a much more limited use API and would probably end up missing out on a lot of value under the time curve, so while I know this API review will be a bit harder, it will have significant positive impact on experience across the ecosystem.